### DISCUSSION TITLE
Understand NextResponse.next()

### DISCUSSION THREAD

**Discussion by Tvgo77** [vercel/next.js#66176] [Help]
### Summary

Didn't expect that response header set in middleware can be accessed in Page component. 

I create a middleware similar to [official doc](https://nextjs.org/docs/app/building-your-application/routing/middleware#setting-headers) to set headers of request and response like this:
```typescript
import { NextRequest, NextResponse } from "next/server"

export function middleware(req: NextRequest) {
    const newHeader = new Headers(req.headers)
    newHeader.set("key1", "val1")

    const response = NextResponse.next({
        request: {
            headers: newHeader
        }
    })
    
    response.headers.set("key2", "val2")
    return response 
}
```
And try to retrieve request header in Page component using [headers()](https://nextjs.org/docs/app/api-reference/functions/headers)
```typescript
import { headers } from "next/headers";

export default function Home() {
  const headerList = headers()

  return (
    <div className="container">
      Test Page
    </div>
  );
}
```
I didn't expect that header "key2" set in response can be accessed in Page component. As my knowledge based on other framework, middleware modifies header of response after response generated by router handler. How on earth both keys set in request header and response header appear in the same place?
![Screenshot_2024-05-24_08_21_34](https://github.com/vercel/next.js/assets/110230961/521bfaf3-718c-4a65-b5fe-45c825dfd01a)

**Comment by kamisone** [comment]
@Tvgo77  did you get to some explanation ?

**Comment by lobre** [comment]
Hey,

I just had the same question as yours and spent some time understanding the code.

> I didn't expect that header "key2" set in response can be accessed in Page component.

### TLDR

I will start by answering this remark. I did not expect that to happen, but the following line in the code is responsible for what you see:

https://github.com/vercel/next.js/blob/84fde91e03918344c5d356986914ab68a5083462/packages/next/src/server/lib/router-utils/resolve-routes.ts#L602

So it is there, but I don't know why.

### Longer explanation of what I understood from the middleware implementation logic in Next.js

I think the middleware concept alongside its documentation are really confusing regarding what belongs to the request and what belongs to the response. Also, in a middleware chain, one middleware function can return a response which will be taken as the request for the next middleware function. So the concepts of requests and responses can rapidly be blurry depending on the context.

Let me explain what I understood from the Next.js situation in regard to those questions. Next.js only has one middleware function that is executed sequentially. In some languages in comparison, such as Go, a middleware has an execution which flows down (request handling) and then back up (response modification). Here, it is "simpler".

This also mean that the Next.js middleware is a "pre" hook that happens before calling the actual node functions. It does not wrap the node function call. Hence, **it cannot act after the function call has been executed**, for example to modify the response headers lastly.

Because of those limitations, it seems that Next.js teams implemented a few "tricks" to still make it a useful thing.

First use-case: as it executes before the node functions, it is able to modify requests. Lets focus on headers as they are usually what is interesting to modify. How would you modify a request header so that your Page component see them in the node function? The clean way to do this (as per what I understood from the code) is to let the middleware return a `NextResponse.next()` instance, and set requests header this way:

```
const newHeader = new Headers(req.headers)
newHeader.set("key1", "val1")

const response = NextResponse.next({
    request: {
        headers: newHeader
    }
})
```

When you do this, you tell Next.js to create a `NextResponse` object, which is an extension of `Response`. By definition, a response is ... well a response. So request headers do not belong in such an object. But you still give it an object as argument with a `request` key.

I think the dev team chose a `Response` object as the return of the middleware because it is handy for situations where you directly want to respond to the end user without processing further. However, as it is also used for when you want to continue the flow to your node functions, they had to be clever.

Underneath, when Next.js sees this `request` key in the object given as argument to `NextResponse.next()`, it will save those headers in a specific internal header (attached to the response object) called `x-middleware-override-headers`.

Later in the flow, after the middleware returns, when Next.js understands that the request flow should not stop here and continue until your node function, it will unwrap that from the response object that was returned from the middleware, and it will inject them as proper request headers that will be part of what your node function receives. This is why you see `key1=val1` in your Page component, as expected.

For reference, this happens there: https://github.com/vercel/next.js/blob/84fde91e03918344c5d356986914ab68a5083462/packages/next/src/server/lib/router-utils/resolve-routes.ts#L564

Also, you might be interested in reading this function where Next.js builds this `x-middleware-override-headers` logic in a custom header on the `NextReponse` object: https://github.com/vercel/next.js/blob/84fde91e03918344c5d356986914ab68a5083462/packages/next/src/server/web/spec-extension/response.ts#L27

Now, to continue on the other outcomes of the middleware, there are a few things that you can do. You can do:
- `NextResponse.redirect(url)`
- `NextResponse.rewrite(url)`
- `Response.json()`

For all those situations, the middleware will break the execution flow and will not forward that to any node functions. Instead, it will generate an actual response and either send it back to the end user, or reboot and re-execute a new request that will go again in the middleware (talking about the `rewrite`).

I think that for those scenarios, the documentation and even the concepts themselves make sense, as they are only related to a response. So basically, this `NextResponse` is returned from your middleware, and Next.js will more or less transform it into an actual response that will get returned to the user. No specific hacks here.

To finish, I will say that I don't have the answer to the "why" of your question quoted as the beginning of this message, that leads to `key2=val2` being present on the node function when executing your Page component.

You are effectively adding this header directly on the `NextResponse` object, meaning that it is meant to be used as a response header for the end user. However, in this example, we are not stopping the flow, as no redirects/rewrites or custom responses are being made. So after the middleware has finished executing, Next.js will take this `NextResponse` object returned, and will continue to the logic that will eventually invoke the node function.

At that moment, and as we said earlier, the Next.js middleware only runs before the node function, and will not run again after. So it cannot modify the actual response headers as returned by the node function. Instead, Next.js will save the response headers defined on the `NextResponse` object returned by the middleware, and will inject them at the end when returning the response to the end user.

But right after the middleware execution, it does:

```
if (value) {
    resHeaders[key] = value
    req.headers[key] = value
}
```

So it puts those headers on the actual object `resHeaders` that will be used for the response for the end user, but it also puts them on the `req` object that is the one that will flow until your node function. And I don't know the reason for that. But effectively, this effect is that response headers put directly on the `NextResponse` will end up both in the end response, but also on the request that flows to your function.

If someone has the reason, I would be glad to hear it!

**Comment by TH-VF** [comment]
Stumbled about the same issue today, very confusing!
